dask/order.py

```diff
@@ -108,108 +109,321 @@ def order(dsk, dependencies=None):
+                deps = dependents[
+                    outer_stack_seeds[outer_stack_seeds_index]
+                ].difference(result)
+                if deps:
+                    if 1 < len(deps) < 1000:
+                        outer_stack_extend(
+                            sorted(deps, key=dependents_key, reverse=True)
+                        )
+                    else:
+                        outer_stack_extend(deps)
+                outer_stack_seeds_index += 1
+
+        if len(dependencies) == len(result):
+            break  # all done!
+
+        # We just finished computing a connected group.
+        # Let's choose the first `item` in the next group to compute.
+        # If we have few large groups left, then it's best to find `item` by taking a minimum.
+        # If we have many small groups left, then it's best to sort.
+        # If we have many tiny groups left, then it's best to simply iterate.
+        if not is_init_sorted:
+            prev_len = len(init_stack)
+            init_stack = init_stack.difference(result)
+            N = len(init_stack)
+            m = prev_len - N
+            # is `min` likely better than `sort`?
+            if m >= N or N + (N - m) * log(N - m) < N * log(N):
+                item = min(init_stack, key=initial_stack_key)
+                continue
+
+            if len(init_stack) < 10000:
+                init_stack = sorted(init_stack, key=initial_stack_key, reverse=True)
+            else:
+                init_stack = list(init_stack)
+            init_stack_pop = init_stack.pop
+            is_init_sorted = True
+
+        item = init_stack_pop()
+        while item in result:
+            item = init_stack_pop()
 
     return result
 
 
-def ndependents(dependencies, dependents, total_dependencies):
-    """ Number of total data elements that depend on key
+def graph_metrics(dependencies, dependents, total_dependencies):
+    """ Useful measures of a graph used by ``dask.order.order``
+Number of total data elements that depend on key
 
-    For each key we return the number of keys that can only be run after this
-    key is run.  The root nodes have value 1 while deep child nodes will have
-    larger values.
-
-    We also return the minimum value of the maximum number of dependencies of
-    all final dependencies (see module-level comment for more)
+    For each key we return:
+    1.  The number of keys that can only be run after this key is run.  The
+        root nodes have value 1 while deep child nodes will have larger values.
+    2.  The minimum value of the maximum number of dependencies of
+        all final dependencies (see module-level comment for more)
+    3.  The maximum value of the maximum number of dependencies of
+        all final dependencies (see module-level comment for more)
+    4.  The minimum height from a root node
+    5.  The maximum height from a root node
```

Comment:
I find the description of 2 and 3 difficult to understand, perhaps an ascii diagram here would be useful to clarify these metrics?

Reply:
docstring updated.  Hopefully better.

Reply:
@jcrist does this look OK now?